File: vscode/src/lang/dart/PubspecDepLensInlayProvider.ts
import * as vscode from "vscode";
import { BaseDepLensInlayProvider } from "../BaseDepLensInlayProvider";
import { PubInlayUtils } from "../../utils/inlay/PubInlayUtils";
import { GithubRepoInfoService } from "../../utils/service/GithubRepoInfoService";
import { PubPkgInfoService } from "../../utils/service/PubPkgInfoService";

const DEP_SECTION_NAMES = new Set([
  "dependencies",
  "dev_dependencies",
  "dependency_overrides",
]);

interface PubspecDependencyEntry {
  line: number;
  name: string;
}

export class PubspecDepLensInlayProvider extends BaseDepLensInlayProvider {
  constructor() {
    super();
    PubPkgInfoService.getInstance().onDidUpdatePackageInfo(() => {
      this.emitter.fire();
    });
    GithubRepoInfoService.getInstance().onDidUpdateRepoInfo(() => {
      this.emitter.fire();
    });
  }

  protected isFileSupported(document: vscode.TextDocument): boolean {
    return document.fileName.endsWith("pubspec.yaml");
  }

  protected async provideInlayHintsForDocument(
    document: vscode.TextDocument,
    range: vscode.Range,
    token: vscode.CancellationToken,
  ): Promise<vscode.InlayHint[]> {
    const hints: vscode.InlayHint[] = [];

    for (const entry of this.collectDependencyEntries(document)) {
      if (token.isCancellationRequested) {
        return hints;
      }

      if (entry.line < range.start.line || entry.line > range.end.line) {
        continue;
      }

      const lineText = document.lineAt(entry.line).text;
      const position = new vscode.Position(entry.line, lineText.length);
      PubInlayUtils.addPubDepInlay(hints, entry.name, position);
    }

    return hints;
  }

  private collectDependencyEntries(
    document: vscode.TextDocument,
  ): PubspecDependencyEntry[] {
    const entries: PubspecDependencyEntry[] = [];
    let sectionIndent: number | undefined;

    for (let line = 0; line < document.lineCount; line++) {
      const text = document.lineAt(line).text;
      const trimmed = text.trim();
      const indent = this.getIndent(text);

      const sectionMatch = text.match(
        /^(\s*)(dependencies|dev_dependencies|dependency_overrides):\s*(?:#.*)?$/,
      );
      if (sectionMatch) {
        if (!DEP_SECTION_NAMES.has(sectionMatch[2])) {
          continue;
        }
        sectionIndent = sectionMatch[1].length;
        continue;
      }

      if (sectionIndent === undefined) {
        continue;
      }

      if (trimmed && !trimmed.startsWith("#") && indent <= sectionIndent) {
        sectionIndent = undefined;
        line--;
        continue;
      }

      if (!trimmed || trimmed.startsWith("#")) {
        continue;
      }

      const depMatch = text.match(/^(\s*)([A-Za-z0-9_]+):\s*(.*)$/);
      if (!depMatch) {
        continue;
      }

      const depIndent = depMatch[1].length;
      if (depIndent !== sectionIndent + 2) {
        continue;
      }

      const name = depMatch[2];
      const value = depMatch[3].trim();

      if (name === "flutter") {
        continue;
      }

      if (this.isUnsupportedInlineSource(value)) {
        continue;
      }

      if (!value && this.hasUnsupportedNestedSource(document, line, depIndent)) {
        continue;
      }

      entries.push({ line, name });
    }

    return entries;
  }

  private hasUnsupportedNestedSource(
    document: vscode.TextDocument,
    startLine: number,
    parentIndent: number,
  ): boolean {
    for (let line = startLine + 1; line < document.lineCount; line++) {
      const text = document.lineAt(line).text;
      const trimmed = text.trim();

      if (!trimmed || trimmed.startsWith("#")) {
        continue;
      }

      const indent = this.getIndent(text);
      if (indent <= parentIndent) {
        break;
      }

      if (/^(sdk|path|git):\b/.test(trimmed)) {
        return true;
      }
    }

    return false;
  }

  private isUnsupportedInlineSource(value: string): boolean {
    return /\b(sdk|path|git)\s*:/.test(value);
  }

  private getIndent(line: string): number {
    return line.length - line.trimStart().length;
  }
}

File: jetbrains/src/main/kotlin/deplens/lang/dart/PubspecDepLensInlayProvider.kt
package deplens.lang.dart

import com.intellij.codeInsight.hints.declarative.InlayTreeSink
import com.intellij.openapi.editor.Document
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import deplens.lang.BaseDepLensInlayProvider
import deplens.utils.inlay.PubInlayUtils

class PubspecDepLensInlayProvider : BaseDepLensInlayProvider() {

    override val usesWholeFileCollection: Boolean = true

    override fun isFileSupported(file: PsiFile): Boolean = file.name == "pubspec.yaml"

    override fun collectElement(file: PsiFile, element: PsiElement, sink: InlayTreeSink) = Unit

    override fun collectFile(file: PsiFile, sink: InlayTreeSink) {
        val document = file.viewProvider.document ?: return
        val entries = collectDependencyEntries(document)

        for ((lineNumber, packageName) in entries) {
            val offset = document.getLineEndOffset(lineNumber)
            PubInlayUtils.addPubDepInlay(file, sink, packageName, offset)
        }
    }

    private fun collectDependencyEntries(document: Document): List<Pair<Int, String>> {
        val entries = mutableListOf<Pair<Int, String>>()
        var sectionIndent: Int? = null

        for (lineNumber in 0 until document.lineCount) {
            val lineText = document.getLineText(lineNumber)
            val trimmed = lineText.trim()
            val indent = lineText.length - lineText.trimStart().length

            val sectionMatch = sectionPattern.matchEntire(lineText)
            if (sectionMatch != null) {
                sectionIndent = sectionMatch.groupValues[1].length
                continue
            }

            val currentSectionIndent = sectionIndent ?: continue

            if (trimmed.isNotEmpty() && !trimmed.startsWith("#") && indent <= currentSectionIndent) {
                sectionIndent = null
                continue
            }

            if (trimmed.isEmpty() || trimmed.startsWith("#")) {
                continue
            }

            val dependencyMatch = dependencyPattern.matchEntire(lineText) ?: continue
            val dependencyIndent = dependencyMatch.groupValues[1].length
            if (dependencyIndent != currentSectionIndent + 2) {
                continue
            }

            val packageName = dependencyMatch.groupValues[2]
            val value = dependencyMatch.groupValues[3].trim()

            if (packageName == "flutter") {
                continue
            }

            if (isUnsupportedInlineSource(value)) {
                continue
            }

            if (value.isEmpty() && hasUnsupportedNestedSource(document, lineNumber, dependencyIndent)) {
                continue
            }

            entries += lineNumber to packageName
        }

        return entries
    }

    private fun hasUnsupportedNestedSource(
        document: Document,
        startLine: Int,
        parentIndent: Int
    ): Boolean {
        for (lineNumber in startLine + 1 until document.lineCount) {
            val lineText = document.getLineText(lineNumber)
            val trimmed = lineText.trim()

            if (trimmed.isEmpty() || trimmed.startsWith("#")) {
                continue
            }

            val indent = lineText.length - lineText.trimStart().length
            if (indent <= parentIndent) {
                break
            }

            if (unsupportedSourcePattern.containsMatchIn(trimmed)) {
                return true
            }
        }

        return false
    }

    private fun isUnsupportedInlineSource(value: String): Boolean {
        return unsupportedSourcePattern.containsMatchIn(value)
    }

    private fun Document.getLineText(lineNumber: Int): String {
        val startOffset = getLineStartOffset(lineNumber)
        val endOffset = getLineEndOffset(lineNumber)
        return getText(TextRange(startOffset, endOffset))
    }

    companion object {
        private val sectionPattern =
            Regex("""^(\s*)(dependencies|dev_dependencies|dependency_overrides):\s*(?:#.*)?$""")
        private val dependencyPattern = Regex("""^(\s*)([A-Za-z0-9_]+):\s*(.*)$""")
        private val unsupportedSourcePattern = Regex("""\b(sdk|path|git)\s*:""")
    }
}
